test/retry.js

```diff
@@ -0,0 +1,199 @@
+                assert.equal(actualThis, expectedThis);
+            });
+        });
+    });
+
+    suite('invocation of "when" function', function() {
+        var expectedThis = {};
+        var expectedArg = {};
+        var actualThis, actualArgs;
+        var action = function() { return expectedArg; };
+        var when = function() {
+            actualThis = this;
+            actualArgs = Array.prototype.slice.call(arguments);
+            return true;
+        };
+        setup(function() {
+            actualThis = null;
+            actualArgs = null;
+        });
+
+        test('first invocation', function() {
+            return retry({
+                maxAttempts: 2,
+                action: action,
+                when: when
+            }).call(expectedThis).then(function() {
+                assert.deepEqual(actualArgs, [expectedArg]);
+                assert.equal(actualThis, expectedThis);
+            });
+        });
+
+        test('subsequent invocations', function() {
+            return retry({
+                maxAttempts: 3,
+                action: action,
+                when: when
+            }).call(expectedThis).then(function() {
+                assert.deepEqual(actualArgs, [expectedArg]);
+                assert.equal(actualThis, expectedThis);
+            });
+        });
+    });
+
+    test('resolution value', function() {
+        var expected = {};
+        var action = function() { return q.resolve(expected); };
+        var when = function() { return false; };
+
+        return retry({
+            maxAttempts: 1,
+            action: action,
+            when: when
+        })().then(function(actual) {
+            assert.equal(actual, expected);
+        });
+    });
+
+    test('exception value', function() {
+        var expected = {};
+        var action = function() { throw expected; };
+        var when = function() { return false; };
+
+        return retry({
+            maxAttempts: 1,
+            action: action,
+            when: when
+        })().then(function() {
+            throw new Error('This promise should not be fulfilled.');
+        }, function(actual) {
+            assert.equal(actual, expected);
+        });
+    });
+
+    test('rejection value', function() {
+        var expected = {};
+        var action = function() { return q.reject(expected); };
+        var when = function() { return false; };
+
+        return retry({
+            maxAttempts: 1,
+            action: action,
+            when: when
+        })().then(function() {
+            throw new Error('This promise should not be fulfilled.');
+        }, function(actual) {
+            assert.equal(actual, expected);
+        });
+    });
+
+    test('retry until maxAttempts', function() {
+        var callCount = 0;
+        var action = function() { callCount += 1; };
+        var when = function() { return true; };
+
+        return retry({
+            maxAttempts: 23,
```

Comment:
Hmm, maybe the tests just have a large number of retries? Do we need this many retries?